File: 2025Bot/Subsystems/SwerveModule.py
import rev
import phoenix6
import math
import wpilib
import commands2
import wpilib.counter
import wpimath.controller
import wpimath.kinematics
import wpimath.units
from wpimath.geometry import Rotation2d
from wpimath.kinematics import SwerveModuleState, SwerveModulePosition


kWheelRadius = 0.0508 #Wheel radius in Meters
kDriveEncoderRes = 2048 #TalonFX Encoder Resolution
kMaxAngularVelocity = math.pi
kMaxAngularAcceleration = math.tau
kGearRatio = 6.75

def enc2Distance(EncoderPosition) -> float: #Converts the current position of the Motor (rotations) into a unit of distance traveled (Meters)
    return ((EncoderPosition / kGearRatio) * (2 * math.pi * kWheelRadius))

def rps2mps(rotations) -> float: #Converts from rotations per second to meters per Second
    return ((rotations * (2 * math.pi * kWheelRadius)) / kGearRatio)

def enc2Rad(value): #converts the encoder value (a range from 0 to 1) to a value in radians.
    radians = (value - 0.5) * 2 *math.pi
    return radians

def setDrivePID(kP: float, kI: float, kD: float):
    wpimath.controller.PIDController(kP, kI, kD)


class swerveModule(commands2.Subsystem):
    def __init__(
        self,
        DriveMotorID: int,
        RotationMotorID: int,
        RotationEncoderPort: int,
        EncoderOffset: float
        ) -> None:

        """
        Initialize a single swerve module containing one TalonFX Drive Motor, one REV Neo Rotation Motor, and one absolute encoder.
        """
        
        #Hardware init
        self.driveMotor = phoenix6.hardware.TalonFX(DriveMotorID)
        self.rotationMotor = rev.SparkMax(RotationMotorID, rev.SparkMax.MotorType.kBrushless)
        self.rotationEncoder = wpilib.AnalogEncoder(RotationEncoderPort, 1, EncoderOffset)

        #Just to make sure that the correct settings are actually applied to the motors.
        motorConfig = phoenix6.configs.TalonFXConfiguration()

        inverted = motorConfig.motor_output.with_inverted(phoenix6.signals.InvertedValue.CLOCKWISE_POSITIVE)
        brake = motorConfig.motor_output.with_neutral_mode(phoenix6.signals.NeutralModeValue.BRAKE)
        currents = motorConfig.current_limits.with_stator_current_limit_enable(True)
        currents.with_stator_current_limit(40)

        motorConfig.with_motor_output(inverted)
        motorConfig.with_current_limits(currents)
        motorConfig.with_motor_output(brake)
        motorConfig.serialize()

        self.driveMotor.configurator.apply(motorConfig)

        rotationConfig = rev.SparkMaxConfig()
        rotationConfig.smartCurrentLimit(40)
        rotationConfig.setIdleMode(rotationConfig.IdleMode.kBrake)
        rotationConfig.inverted(True)

        self.rotationMotor.configure(rotationConfig, self.rotationMotor.ResetMode.kResetSafeParameters, self.rotationMotor.PersistMode.kNoPersistParameters)

        
        #Feed Forward Control
        self.rotationMotorFeedForward = wpimath.controller.SimpleMotorFeedforwardMeters(1, 0.5)


        super().__init__()

    def setDrivePID(self, kP: float, kI: float, kD: float):
        self.drivePIDController = wpimath.controller.PIDController(
            kP,  # Proportional gain
            kI,   # Integral gain
            kD,   # Derivative gain
            
        )

    def setDriveFF(self, kS: float, kV: float, kA: float):
        self.driveMotorFeedForward = wpimath.controller.SimpleMotorFeedforwardMeters(kS, kV, kA)
    
    def setRotationPID(self, kP: float, kI: float, kD: float):
        self.rotationPIDController = wpimath.controller.PIDController(
            kP,  # Proportional gain
            kI,   # Integral gain
            kD,   # Derivative gain
        )

        self.rotationPIDController.enableContinuousInput(-math.pi, math.pi)
        self.rotationPIDController.setSetpoint(0.0)
        
    def getState(self):
        """
        Gets the current state of a single swerve module (Both Drive and Rotation motors)
        """
        return SwerveModuleState(
            rps2mps(self.driveMotor.get_velocity().value_as_double), #Gets the speed of the wheels in m/s
            Rotation2d(enc2Rad(self.rotationEncoder.get())) #Converts the position into radians as rotation2d requests
        )        
    
    def getPosition(self):
        """
        Gets the current position of a swerve module (both the drive and rotation motors)
        """
        return SwerveModulePosition(
            enc2Distance(self.driveMotor.get_position().value_as_double), #gets the current position of the wheels
            Rotation2d(enc2Rad(self.rotationEncoder.get())) #Converts the position into radians as rotation2d requests
        )
    
    def setState(
            self,
            newState: SwerveModuleState
    ) -> None:
        """
        Sets a new state for the swerve module to move to.
        """
        newState.optimize(Rotation2d(enc2Rad(self.rotationEncoder.get())))
        newState.cosineScale(Rotation2d(enc2Rad(self.rotationEncoder.get())))

        driveOutput = self.drivePIDController.calculate(rps2mps(self.driveMotor.get_velocity().value_as_double), newState.speed)
        driveFF = self.driveMotorFeedForward.calculate(newState.speed)

        rotationOutput = self.rotationPIDController.calculate(enc2Rad(self.rotationEncoder.get()), newState.angle.radians())

        #self.driveMotor.setVoltage((newState.speed / 4.72) * 13)
        self.driveMotor.setVoltage(driveOutput + driveFF)
        self.rotationMotor.set(rotationOutput)

    def stopAllMotors(self):
        self.driveMotor.stopMotor()
        self.rotationMotor.stopMotor()

File: JavaExamples/SwerveJava/src/main/java/frc/robot/subsystems/SwerveModule.java
package frc.robot.subsystems;

import com.revrobotics.spark.SparkMax;
import com.revrobotics.spark.config.SparkMaxConfig;
import com.revrobotics.spark.config.SparkBaseConfig.IdleMode;
import com.revrobotics.spark.SparkBase.PersistMode;
import com.revrobotics.spark.SparkBase.ResetMode;
import com.revrobotics.spark.SparkLowLevel.MotorType;

import com.ctre.phoenix6.hardware.TalonFX;
import com.ctre.phoenix6.configs.TalonFXConfiguration;
import com.ctre.phoenix6.signals.InvertedValue;
import com.ctre.phoenix6.signals.NeutralModeValue;

import edu.wpi.first.wpilibj2.command.SubsystemBase;
import edu.wpi.first.math.geometry.Rotation2d;
import edu.wpi.first.math.kinematics.SwerveModulePosition;
import edu.wpi.first.math.kinematics.SwerveModuleState;
import edu.wpi.first.wpilibj.AnalogEncoder;
import edu.wpi.first.math.controller.PIDController;

import frc.robot.Constants.RobotConstants;

public class SwerveModule extends SubsystemBase {
    /* 
     * Drive Motor: TalonFX (Kraken X60)
     * Rotation Motor: SparkMax (Rev NEO)
     * Rotation Encoder: Analog Encoder (ThrifyBot Absolute Magnetic Encoder)
     */
    
    private final TalonFX m_driveMotor;
    private final SparkMax m_rotationMotor;
    private final AnalogEncoder m_rotationEncoder;

    /* 
     * PID Controller for the rotation motor
     * The PID controller is used to control the rotation of the module
     */
    private final PIDController m_rotationPIDController = new PIDController(0, 0, 0);

    /// Conversion equations
    private double enc2Distance(double EncoderPosition){
        return ((EncoderPosition / RobotConstants.kGearRatio) * (2 * Math.PI * RobotConstants.kWheelRadius));
    }

    private double rps2mps(double EncoderRotations){
        return ((EncoderRotations * (2 * Math.PI * RobotConstants.kWheelRadius)) / RobotConstants.kGearRatio);
    }

    private double enc2Rad(double EncoderPosition){
        return ((EncoderPosition - 0.5) * 2 * Math.PI);
    }
    
    //Constructor
    public SwerveModule(
        int DriveMotorID,
        int RotationMotorID,
        int RotationEncoderID,
        double RotationEncoderOffset) {
        m_driveMotor = new TalonFX(DriveMotorID);
        m_rotationMotor = new SparkMax(RotationMotorID, MotorType.kBrushless);
        m_rotationEncoder = new AnalogEncoder(RotationEncoderID, 0, RotationEncoderOffset);

        /* 
         * Drive Motor Configuration
         * Rotation Motor Configuration
         */
        TalonFXConfiguration driveConfig = new TalonFXConfiguration();

        driveConfig.MotorOutput.withInverted(InvertedValue.CounterClockwise_Positive);
        driveConfig.MotorOutput.withNeutralMode(NeutralModeValue.Brake);
        driveConfig.CurrentLimits.withStatorCurrentLimit(40);
        driveConfig.CurrentLimits.withStatorCurrentLimitEnable(true);
        driveConfig.serialize();

        SparkMaxConfig rotationConfig = new SparkMaxConfig();
        
        rotationConfig
            .smartCurrentLimit(40)
            .idleMode(IdleMode.kBrake);

        m_driveMotor.getConfigurator().apply(driveConfig);
        m_rotationMotor.configure(rotationConfig, ResetMode.kResetSafeParameters, PersistMode.kPersistParameters);
        

        m_rotationPIDController.enableContinuousInput(-Math.PI, Math.PI);        
    }

    /* 
     * Set the PID values for the rotation motor
     * kP: Proportional gain
     * kI: Integral gain
     * kD: Derivative gain
     */
    public void setPIDValues(double kP, double kI, double kD) { 
        m_rotationPIDController.setPID(kP, kI, kD);
    }

    /* 
     * Get the current state of the module
     * The state includes the speed and angle of the module
     * The speed is in meters per second and the angle is in radians
     * The speed is calculated using the encoder velocity, gear ratio, and wheel radius.
     * The angle is calculated using the encoder position and the gear ratio
     */
    public SwerveModuleState getState(){
        return new SwerveModuleState(
            rps2mps(m_driveMotor.getVelocity().getValueAsDouble()),
            new Rotation2d(enc2Rad(m_rotationEncoder.get()))
        );
    }

    /* 
     * Get the current position of the module
     * The position includes the distance and angle of the module
     * The distance is in meters and the angle is in radians
     * The distance is calculated using the encoder position, gear ratio, and wheel radius.
     * The angle is calculated using the encoder position and the gear ratio
     */
    public SwerveModulePosition getPosition(){
        return new SwerveModulePosition(
            enc2Distance(m_driveMotor.getPosition().getValueAsDouble()),
            new Rotation2d(enc2Rad(m_rotationEncoder.get()))
        );
    }

    // Set the state of the module
    public void setState(SwerveModuleState newState){
        newState.optimize(new Rotation2d(enc2Rad(m_rotationEncoder.get())));
        newState.cosineScale(new Rotation2d(enc2Rad(m_rotationEncoder.get())));

        double rotationOutput = m_rotationPIDController.calculate(enc2Rad(m_rotationEncoder.get()), newState.angle.getRadians());

        m_driveMotor.setVoltage((newState.speedMetersPerSecond / RobotConstants.kMaxModuleSpeed) / RobotConstants.kMaxVoltage);
        m_rotationMotor.set(rotationOutput);

    }

    // stops the drive and rotation motors
    // This is used when the robot is not moving
    public void stopDrive(){
        m_driveMotor.stopMotor();
        m_rotationMotor.stopMotor();
    }
}
